concurrent_map.go

```diff
@@ -302,6 +348,40 @@ func (m ConcurrentMap) MarshalJSON() ([]byte, error) {
 	return json.Marshal(tmp)
 }
 
+// isEvictionOccurred check if the map size overflow the maximum allowed size.
+func (m ConcurrentMap) isEvictionOccurred(key string) bool {
+	shard := m.GetShard(key)
+	if m.Count() > shard.maxSize {
+		return true
+	}
+	return false
+}
+
+// removeFirstElement removes the first element from the key slice.
+func (m ConcurrentMap) removeFirstElement(key string) {
+	shard := m.GetShard(key)
+	shard.Lock()
+	copy(shard.mapKeys[0:], shard.mapKeys[1:])
+	shard.mapKeys = shard.mapKeys[:len(shard.mapKeys)-1]
+	shard.Unlock()
+}
+
+// removeMapKey removes the map key from the key slice.
+func (m ConcurrentMap) removeMapKey(key string) {
+	shard := m.GetShard(key)
+
+	if len(shard.mapKeys) > 0 {
```

Comment:
this should be under lock

Reply:
As stated on the comment below the shard is locked in the method which invoke `removeMapKey`. It cannot be locked on multiple places, since the goroutine will block until the mutex will be available.